include/boost/geometry/algorithms/detail/intersection/gc.hpp

```diff
@@ -0,0 +1,327 @@
+// Boost.Geometry
+
+// Copyright (c) 2022, Oracle and/or its affiliates.
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_DETAIL_INTERSECTION_GC_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_DETAIL_INTERSECTION_GC_HPP
+
+
+#include <tuple>
+
+#include <boost/geometry/algorithms/detail/intersection/interface.hpp>
+#include <boost/geometry/algorithms/detail/make_rtree.hpp>
+#include <boost/geometry/views/detail/geometry_collection_view.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace intersection
+{
+
+
+template <typename GC, typename Multi>
+struct gc_can_move_element
+{
+    template <typename G>
+    using is_same_as_single = std::is_same<G, typename boost::range_value<Multi>::type>;
+    using gc_types = typename traits::geometry_types<GC>::type;
+    using found_type = typename util::sequence_find_if<gc_types, is_same_as_single>::type;
+    static const bool value = ! std::is_void<found_type>::value;
+};
+
+template <typename GC, typename Multi>
+struct gc_can_convert_element
+{
+    template <typename G>
+    using has_same_tag_as_single = std::is_same
+        <
+            typename geometry::tag<G>::type,
+            typename geometry::tag<typename boost::range_value<Multi>::type>::type
+        >;
+    using gc_types = typename traits::geometry_types<GC>::type;
+    using found_type = typename util::sequence_find_if<gc_types, has_same_tag_as_single>::type;
+    static const bool value = ! std::is_void<found_type>::value;
+};
+
+template
+<
+    typename GC, typename Multi,
+    std::enable_if_t<gc_can_move_element<GC, Multi>::value, int> = 0
+>
+inline void gc_move_multi_back_(GC& gc, Multi&& multi)
+{
+    range::emplace_back(gc, std::move(*boost::begin(multi)));
+}
+
+template
+<
+    typename GC, typename Multi,
+    std::enable_if_t<! gc_can_move_element<GC, Multi>::value && gc_can_convert_element<GC, Multi>::value, int> = 0
+>
+inline void gc_move_multi_back_(GC& gc, Multi&& multi)
+{
+    typename gc_can_convert_element<GC, Multi>::found_type single_out;
+    geometry::convert(*boost::begin(multi), single_out);
+    range::emplace_back(gc, std::move(single_out));
+}
+
+template
+<
+    typename GC, typename Multi,
+    std::enable_if_t<! gc_can_move_element<GC, Multi>::value && ! gc_can_convert_element<GC, Multi>::value, int> = 0
```

Comment:
So we use `enable_if` here i/o tag dispatching because it's not really based on geometry tags, but more on just some conditions. That make sense to me for these cases.